Add support for custom checkbox to wxInfoBar

Blake-Madden opened the pull request:
For the generic `wxInfoBar`, adds a feature to show a custom checkbox at the bottom.

By intercepting the `wxID_CLOSE` button event, the client can check the state of this checkbox and then know whether they should show it again or not.

![image](https://github.com/user-attachments/assets/d5ecba35-b917-4e0d-8fbf-4a7dd42fd2d4)

The inclusion of this checkbox can be toggled and is not shown by default.

Blake-Madden committed:
Add support for "Do not show this again" checkbox to wxInfoBar

Blake-Madden committed:
Fix spaces

Blake-Madden committed:
Fix forward declare

vadz commented on include/wx/generic/infobar.h:
```diff
@@ -75,6 +75,21 @@ class WXDLLIMPEXP_CORE wxInfoBarGeneric : public wxInfoBarBase
+    bool IsIncludingDontShowAgainCheckbox() const { return m_includeDontShowAgain; }
+
+    // Sets whether the "Do not show this again" checkbox should be shown.
+    void IncludeDontShowAgainCheckbox(const bool includeCheckbox)
```
Top-level const is ignored and shouldn't be used.
```suggestion
    void IncludeDontShowAgainCheckbox(bool includeCheckbox)
```

vadz commented on include/wx/generic/infobar.h:
```diff
@@ -75,6 +75,21 @@ class WXDLLIMPEXP_CORE wxInfoBarGeneric : public wxInfoBarBase
+
+    // Sets whether the "Do not show this again" checkbox should be shown.
+    void IncludeDontShowAgainCheckbox(const bool includeCheckbox)
+    {
```
This function shouldn't be inline, there is really no reason to put it in the header.

Blake-Madden replied:
Yes, I see the compile error now. Will fix.

vadz commented on src/generic/infobar.cpp:
```diff
@@ -115,19 +117,37 @@ bool wxInfoBarGeneric::Create(wxWindow *parent, wxWindowID winid)
     m_button->SetToolTip(_("Hide this notification message."));
 
+    m_dontShowAgainCheckbox =
+        new wxCheckBox(this, wxID_ANY, _(L"Do not show this again."), wxDefaultPosition,
```
Why `L`?
```suggestion
        new wxCheckBox(this, wxID_ANY, _("Do not show this again."), wxDefaultPosition,
```

Blake-Madden replied:
Sorry, force of habit. I use `L` in my code usually, but not in wx PRs.

vadz reviewed (dismissed):
Thanks, I agree that this could be useful, but I think it would make more sense to

1. Not show the checkbox by default.
2. Add `AddCheckbox()` or, if we want to be consistent with `wxRichMessageDialogBase`, `ShowCheckBox()` function to show it.

Importantly, GTK version should check for the checkbox being shown and fall back to the generic one if it is.

Blake-Madden committed:
Remove const

Co-authored-by: VZ <vz-github@zeitlins.org>

Blake-Madden committed:
Remove L

Co-authored-by: VZ <vz-github@zeitlins.org>

Blake-Madden committed:
Move function to implemenation file

Blake-Madden committed:
Redesign API to be like wxRichMessageDialog

Blake-Madden committed:
Fix include file

Blake-Madden commented:
>     1. Not show the checkbox by default.

Yes, it is hidden by default.

>     2. Add `AddCheckbox()` or, if we want to be consistent with `wxRichMessageDialogBase`, `ShowCheckBox()` function to show it.

Got it. Now uses the same API as `wxRichMessageDialogBase`.

> Importantly, GTK version should check for the checkbox being shown and fall back to the generic one if it is.

Not sure how to approach this because `wxInfoBar::Create` constructs the control either generically or natively. At that point, `ShowCheckBox()` couldn't have been called yet (that would have to come later) and by then it is too late. I don't see a way to force it to be generic from `ShowCheckBox()` because by then it is already created natively.

I could add an warning in the GTK version of `ShowCheckBox()` if it is running natively, saying that you should use the generic version.

Blake-Madden committed:
Use quotes, not angles for include file

vadz commented:
To be able to decide which version to use we'd need to use a style (styles are available in `Create()`). I'm not sure if it's worth the extra complexity, but forcing the user to use `wxInfoBarGeneric` with wxGTK is not ideal, it would be really nice to provide some way to use just `wxInfoBar` everywhere.

Blake-Madden committed:
Add style flag to infobar to force use of generic version
Default is this to be turned off to preserve backward compatability

Blake-Madden commented:
I added a new window style `wxINFOBAR_GENERIC` to control this. I'm going to do some more testing on Linux and then change it back from draft...

Blake-Madden committed:
Expand docs

vadz commented:
> I added a new window style `wxINFOBAR_GENERIC` to control this.

Calling the style `wxINFOBAR_CHECKBOX` would be better, like this we could switch to using native version on GTK (or some other platform) in the future if it adds support for this — while with the current style we'd always have to use the generic one.

Blake-Madden committed:
Revert "Expand docs"

This reverts commit b90363cd823e7a1cf449a27deaab5c0beaeaeb69.

Blake-Madden committed:
Revert "Add style flag to infobar to force use of generic version"

This reverts commit 231a6fcf70708c05f8faac06fcc1a7d480908a2e.

Blake-Madden committed:
Add wxINFOBAR_CHECKBOX

Blake-Madden commented:
OK, just added `wxINFOBAR_CHECKBOX`. It is not on by default, but if you call `ShowCheckBox()` without this style you will get a "helpful" assert message.

vadz commented on include/wx/generic/infobar.h:
```diff
@@ -96,10 +110,14 @@ class WXDLLIMPEXP_CORE wxInfoBarGeneric : public wxInfoBarBase
     // should be called when we're shown or hidden)
     void UpdateParent();
 
+    long m_style = 0;
```
I don't think we need this one, we should store the style in `wxWindow::m_windowStyle`.

vadz reviewed (commented):
Thanks, looks good now, I'd just like to remove `m_style` because it's confusing to have both it and `m_windowStyle`.

Of course, please let me know if it's really needed for some reason I'm missing. Thanks!

Blake-Madden committed:
Store value in m_windowStyle

vadz reviewed (commented):
Sorry, I've realized there is something else I forgot about: it would be nice to add the new style to `wxInfoBarXmlHandler` in `src/xrc/xh_infobar.cpp` (and also to `docs/doxygen/overviews/xrc_format.h`). Could you please do it? TIA!

Blake-Madden committed:
Improve function description

Blake-Madden committed:
Add support to XRC handler

Blake-Madden committed:
Update src/generic/infobar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

Blake-Madden committed:
Update src/gtk/infobar.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

Blake-Madden commented:
OK, I added the style and parameters to `ShowCheckbox()` via XRC.

vadz committed:
Rename "checkboxvalue" to "checkboxlabel"

Also enable wxINFOBAR_CHECKBOX automatically when it is specified.

And complain if "checked" is specified but ignored.

vadz committed:
Add new wxInfoBar attributes to the schema

vadz commented:
Thanks! I was going to merge this but realized I forgot to ask you to add the new attributes to the schema, and so started doing it, but then realized that `checkboxvalue` was rather inconsistent with the similar attributes names and so renamed it to `checkboxlabel` and also changed its semantics slightly: it now always enables `wxINFOBAR_CHECKBOX` because there just doesn't seem any reason not to do it.

I've pushed my changes into your branch, please let me know if you have any remarks, otherwise I'll merge this soon.

Blake-Madden commented:
Sounds good to me, thanks!

Yeah, I wasn't sure if it was OK to enable that by default since it would cause it to go generic under GTK, but that seems OK to me.
